src/record-step-state-access.h

```diff
@@ -0,0 +1,840 @@
+            return; // already saved
+        }
+        auto [it, _] = m_touched_pages.emplace(page, page_data_type());
+        m_m.read_memory(page, it->second.data(), it->second.size());
+    }
+
+    /// \brief Get the sibling hashes of all touched pages
+    sibling_hashes_type get_sibling_hashes() {
+        sibling_hashes_type sibling_hashes{};
+        // page address are converted to page indices, in order to avoid overflows
+        page_indices_type page_indices{};
+        // iterate in ascending order of page addresses (the container is ordered by key)
+        for (const auto &[address, _] : m_touched_pages) {
+            page_indices.push_back(address >> LOG2_PAGE_SIZE);
+        }
+        auto next_page_index = page_indices.cbegin();
+        get_sibling_hashes_impl(0, LOG2_ROOT_SIZE - LOG2_PAGE_SIZE, page_indices, next_page_index, sibling_hashes);
+        if (next_page_index != page_indices.cend()) {
+            throw std::runtime_error("get_sibling_hashes failed to consume all pages");
+        }
+        return sibling_hashes;
+    }
+
+    /// \brief Recursively get the sibling hashes of all touched pages
+    /// \param page_index index of 1st page in range
+    /// \param page_count_log2_size log2 of the number of pages in range
+    /// \param page_indices indices of all pages
+    /// \param next_page_index smallest page index not visited yet
+    /// \param sibling_hashes stores the collected sibling hashes during the recursion
+    void get_sibling_hashes_impl(address_type page_index, int page_count_log2_size, page_indices_type &page_indices,
+        page_indices_type::const_iterator &next_page_index, sibling_hashes_type &sibling_hashes) {
+        auto page_count = UINT64_C(1) << page_count_log2_size;
+        if (next_page_index == page_indices.cend() || page_index + page_count <= *next_page_index) {
+            // we can skip the merkle tree update, because a full update was done before the recording started
+            sibling_hashes.push_back(m_m.get_merkle_tree_node_hash(page_index << LOG2_PAGE_SIZE,
+                page_count_log2_size + LOG2_PAGE_SIZE, skip_merkle_tree_update));
+        } else if (page_count_log2_size > 0) {
+            get_sibling_hashes_impl(page_index, page_count_log2_size - 1, page_indices, next_page_index,
+                sibling_hashes);
+            get_sibling_hashes_impl(page_index + (UINT64_C(1) << (page_count_log2_size - 1)), page_count_log2_size - 1,
+                page_indices, next_page_index, sibling_hashes);
+        } else {
+            ++next_page_index;
+        }
+    }
+
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+    void do_push_bracket(bracket_type type, const char *text) {
+        (void) type;
+        (void) text;
+    }
+
+    int do_make_scoped_note(const char *text) { // NOLINT(readability-convert-member-functions-to-static)
+        (void) text;
+        return 0;
+    }
+
+    uint64_t do_read_x(int reg) const {
```

Comment:
Had a conversation with @mpernambuco and I think we should replace all of these with `do_read_reg()`/`do_write_reg()`. Then, we implement all `i_state_access::do_read_*.*/do_write*.*` on top of `i_state_access::read_reg()/write_reg()`, and we forward `i_state_access::read_reg()/write_reg()` to `i_state_access::do_read_reg()/do_write_reg()` and implement `i_state_access::do_read_reg()/do_write_reg()` to throw and exception. This way, every state access can choose to implement `do_read_*.*()/do_write_*.*()` or only implement `do_read_reg()` and `do_write_reg()`, or even mix both. What do you think, @edubart? We should also merge the enums `shadow_state_reg` with `machine::reg`, so we don't need to convert between them anymore.

Reply:
Maybe do this in a separate PR?

Reply:
I agree, I think this would a good to simplify all state accessors, but indeed this refactoring deserves a separate PR.